Cython/Compiler/Optimize.py

```diff
@@ -228,6 +228,12 @@ def _optimise_for_loop(self, node, iterable, reversed=False):
             return self._transform_bytes_iteration(node, iterable, reversed=reversed)
         if iterable.type is Builtin.unicode_type:
             return self._transform_unicode_iteration(node, iterable, reversed=reversed)
+        if iterable.type in (Builtin.list_type, Builtin.tuple_type):
+            return self._transform_indexable_iteration(node, iterable, reversed=reversed)
+        if isinstance(iterable, ExprNodes.CoerceToPyTypeNode) and iterable.arg.type.is_memoryviewslice:
+            # TODO would it make sense to prevent the coercion? Or does it need to be there since
+            # this is only an optimization stage (so it should work without this set)
```

Comment:
If both versions can work, then it's better to keep this a pure optimisation and not rely on it. It's very reasonable for an optimisation to unpack coerced values.